### PR TITLE
[clang-format] Add Options to break inside the TableGen DAGArg.

### PR DESCRIPTION
**Pull request by hnakamura5**
This adds two options to control the line break inside TableGen DAGArg.
- TableGenBreakInsideDAGArg
- TableGenBreakingDAGArgOperators

### CODE REVIEW

**Comment by hnakamura5** [comment]
These options have a dependency that TableGenBreakInsideDAGArgList is effective only when TableGenBreakingDAGArgOperators is specified as true.
I'm not sure this is a smart way.

File: clang/lib/Format/ContinuationIndenter.cpp [hunk 1]
```diff
@@ -1842,6 +1846,19 @@ void ContinuationIndenter::moveStatePastScopeOpener(LineState &State,
         Style.ContinuationIndentWidth +
         std::max(CurrentState.LastSpace, CurrentState.StartOfFunctionCall);
 
+    if (Style.isTableGen()) {
+      if (Current.is(TT_TableGenDAGArgOpenerToBreak) &&
+          Style.TableGenBreakInsideDAGArg == FormatStyle::DAS_BreakElements) {
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/ContinuationIndenter.cpp:1851]:
```suggestion
    if (Style.isTableGen() && Current.is(TT_TableGenDAGArgOpenerToBreak) &&
          Style.TableGenBreakInsideDAGArg == FormatStyle::DAS_BreakElements) {
```

File: clang/unittests/Format/FormatTestTableGen.cpp [hunk 2]
```diff
@@ -332,6 +332,84 @@ TEST_F(FormatTestTableGen, Assert) {
   verifyFormat("assert !le(DefVar1, 0), \"Assert1\";\n");
 }
 
+TEST_F(FormatTestTableGen, DAGArgBreakElements) {
+  FormatStyle Style = getGoogleStyle(FormatStyle::LK_TableGen);
+  Style.ColumnLimit = 60;
+  // By default, the DAGArg does not have a break inside.
+  verifyFormat("def Def : Parent {\n"
```
Comment (HazardyKnusperkeks) [inline clang/unittests/Format/FormatTestTableGen.cpp:339]:
```suggestion
  ASSERT_EQ(Style.TableGenBreakInsideDAGArg, FormatStyle::DAS_DontBreak);
  verifyFormat("def Def : Parent {\n"
```

File: clang/unittests/Format/TokenAnnotatorTest.cpp [hunk 3]
```diff
@@ -2332,6 +2332,77 @@ TEST_F(TokenAnnotatorTest, UnderstandTableGenTokens) {
...
+  EXPECT_TOKEN(Tokens[1], tok::identifier,
+               TT_TableGenDAGArgOperatorToBreak); // ins
+  EXPECT_TOKEN(Tokens[5], tok::comma, TT_TableGenDAGArgListCommaToBreak);
+  EXPECT_TOKEN(Tokens[9], tok::r_paren, TT_TableGenDAGArgCloser);
+
+  Tokens = AnnotateValue("(other type1:$src1, type2:$src2)");
+  ASSERT_EQ(Tokens.size(), 10u) << Tokens;
+  EXPECT_TOKEN(Tokens[0], tok::l_paren, TT_TableGenDAGArgOpener);
+  EXPECT_TOKEN(Tokens[1], tok::identifier,
+               TT_Unknown); // other
```
Comment (HazardyKnusperkeks) [inline clang/unittests/Format/TokenAnnotatorTest.cpp:2403]:
```suggestion
  EXPECT_TOKEN(Tokens[1], tok::identifier, TT_Unknown); // other
```
This should fit?

Comment (hnakamura5) [reply clang/unittests/Format/TokenAnnotatorTest.cpp]: Fixed as suggested.

File: clang/unittests/Format/FormatTestTableGen.cpp (hunk 2 above)
Comment (hnakamura5) [reply clang/unittests/Format/FormatTestTableGen.cpp:340]: Added the suggested check of default option.

File: clang/lib/Format/ContinuationIndenter.cpp (hunk 1 above)
Comment (hnakamura5) [reply clang/lib/Format/ContinuationIndenter.cpp]: Changed as suggested.

**Review by HazardyKnusperkeks — APPROVED** [review]

**Comment by hnakamura5** [comment]
Thank you very much!
I really appreciate you for reviewing up to such a complicated option.

**Comment by wangpc-pp** [comment]
This breaks CI `Test documentation build` like: https://github.com/llvm/llvm-project/actions/runs/8339765845/job/22822367034

**Comment by HazardyKnusperkeks** [comment]
Presumably fixed by #85760

**Comment by hnakamura5** [comment]
@wangpc-pp 
Thank you for telling, and sorry for overlooking the detailed check for CI.

**Comment by hnakamura5** [comment]
@HazardyKnusperkeks 
Thank you for reviewing and merging the fix.

**Pull request merged**